### ISSUE TITLE
Incorrect return value from Product Attribute Repository

### ISSUE THREAD

**Issue by samtay** [magento/magento2#4803]
To adhere to `Magento\Catalog\Api\ProductAttributeRepositoryInterface` (extends `Magento\Framework\Api\MetadataServiceInterface`), the

``` php
Magento\Catalog\Model\Product\Attribute\Repository::getCustomAttributesMetadata
```

implementation should return `Magento\Framework\Api\MetadataObjectInterface[]`, however it instead returns `Magento\Catalog\Api\Data\ProductAttributeInterface[]`.

The `ProductAttributeInterface` **does** declare the two methods declared in `MetadataObjectInterface` , however they are not related by PHP `extend`.

**Comment by andimov** [comment]
@SamTay, to allow us research the issue, please, provide the used version and description according to the template.

**Comment by samtay** [comment]
The issue exists on all versions, on CE develop branch. The description is in the initial comment. I don't see what more information you need? This isn't a huge deal, I'm just letting you know there is a disconnect in the hierarchy of the interfaces. The documentation is incorrect -- which the framework can be _very_ finicky about sometimes when generating code or parsing objects in the webapi. Something for you to be aware of.

**Comment by vkublytskyi** [comment]
@SamTay, thank you for reporting this issue. To fix this issue we will add declaration that `Magento\Eav\Api\Data\AttributeInterface` extends `Magento\Framework\Api\MetadataObjectInterface`.
Internal ticket `MAGETWO-55438`.

**Comment by magento-engcom-team** [comment]
@samtay, thank you for your report.
We've created internal ticket(s) MAGETWO-55438 to track progress on the issue.

**Comment by julian-van-drielen** [comment]
I will be working on this issue

**Comment by magento-engcom-team** [comment]
Hi @samtay. Thank you for your report. 
The issue has been fixed in magento/magento2#15691 by @cream-julian in 2.2-develop branch
Related commit(s):
 - [4c1989ee90e2cde19cb71445f4c07dd19f8f7b3a](https://github.com/magento/magento2/commit/4c1989ee90e2cde19cb71445f4c07dd19f8f7b3a)
 - [215be5a9d10ec81d243b3a2225c45264632176ef](https://github.com/magento/magento2/commit/215be5a9d10ec81d243b3a2225c45264632176ef)
 - [2639d0ec3ae7b0b2952f9fa1fd3f0f2d73f761c1](https://github.com/magento/magento2/commit/2639d0ec3ae7b0b2952f9fa1fd3f0f2d73f761c1)
 - [43dd85edfb856bca5fb3fdd6d11cdaec00cc96eb](https://github.com/magento/magento2/commit/43dd85edfb856bca5fb3fdd6d11cdaec00cc96eb)

The fix will be available with the upcoming 2.2.7 release.

**Comment by sidolov** [comment]
Hi @samtay. Thank you for your report. 
The issue has been fixed in magento/magento2#15688 by @cream-julian in 2.1-develop branch
Related commit(s):
 - [d409248d850d2118370e1d8f102c60dda10f3e64](https://github.com/magento/magento2/commit/d409248d850d2118370e1d8f102c60dda10f3e64)
 - [90f37dce6c35d17537ed4efc73e3bbb3b8901967](https://github.com/magento/magento2/commit/90f37dce6c35d17537ed4efc73e3bbb3b8901967)
 - [b064cd77d9f763d0e997af93619840836f9efb70](https://github.com/magento/magento2/commit/b064cd77d9f763d0e997af93619840836f9efb70)

The fix will be available with the upcoming 2.1.16 release.

**Comment by magento-engcom-team** [comment]
Hi @samtay. Thank you for your report. 
The issue has been fixed in magento/magento2#15723 by @cream-julian in 2.3-develop branch
Related commit(s):
 - [19ca0905ae81d9d32ae5aa1f56e38b31b1bf7cd0](https://github.com/magento/magento2/commit/19ca0905ae81d9d32ae5aa1f56e38b31b1bf7cd0)
 - [4e51e9c056f29318564dd50ce293a9cedc84bbf1](https://github.com/magento/magento2/commit/4e51e9c056f29318564dd50ce293a9cedc84bbf1)
 - [a8dc518ecd77b813ff79e8fb864c9cb4b01456e9](https://github.com/magento/magento2/commit/a8dc518ecd77b813ff79e8fb864c9cb4b01456e9)
 - [40f046ef08df02f4c145e9bbe2ec11ac9eb47ae9](https://github.com/magento/magento2/commit/40f046ef08df02f4c145e9bbe2ec11ac9eb47ae9)

The fix will be available with the upcoming 2.3.0 release.
